### ISSUE TITLE
Bug: some ks_solver failed in a test case

### ISSUE THREAD

**Issue by dzzz2001** [deepmodeling/abacus-develop#5359]
### Describe the bug

I am testing a si216 example. When setting `ks_solver` to `scalapack_gvx` or `cusolver`, ABACUS fails to run. Only when `ks_solver` is set to `genelpa` does it successfully execute.

Error messages are as follows:![Image](https://github.com/user-attachments/assets/30d0dbfa-bc25-4827-8c27-acb1146406d9)
![Image](https://github.com/user-attachments/assets/951db30f-baee-4fe1-a639-3e503edf0347)

`genelpa` run successfully: ![Image](https://github.com/user-attachments/assets/fccd2d87-4ea2-4bb8-b98f-95622b70630b)

### To Reproduce

1. my ABACUS version is 3.8.1, commit: 7a29e2f56.
2. run ABACUS with “OMP_NUM_THREADS=5, mpirun -n 5
3. test case:

### Task list for Issue attackers (only for developers)

- [ ] Verify the issue is not a duplicate.
- [ ] Describe the bug.
- [ ] Steps to reproduce.
- [ ] Expected behavior.
- [ ] Error message.
- [ ] Environment details.
- [ ] Additional context.
- [ ] Assign a priority level (low, medium, high, urgent).
- [ ] Assign the issue to a team member.
- [ ] Label the issue with relevant tags.
- [ ] Identify possible related issues.
- [ ] Create a unit test or automated test to reproduce the bug (if applicable).
- [ ] Fix the bug.
- [ ] Test the fix.
- [ ] Update documentation (if necessary).
- [ ] Close the issue and inform the reporter (if applicable).

**Comment by kirk0830** [comment]
@dzzz2001 I doubt this is because we use scalapack without preconditioning (the most simple one is Cholesky decomposition, genelpa does, see:
https://github.com/deepmodeling/abacus-develop/blob/7a29e2f56da86fde06b8c052390fc8d580b55801/source/module_hsolver/genelpa/elpa_new_complex.cpp#L156
https://manpages.ubuntu.com/manpages/lunar/man3/pzpotrf.3.html
) on diagnalization problem, plus for this case the overlap matrix S is suspected to be ill-conditioned (such a big cell)...could you try to diagonalize solely the S matrix, and print the ratio between max and min eigen val of it? you can run on single processor so that you do not need to care anything about 2D Block Cyclic Distribution technique.

**Comment by dzzz2001** [comment]
the condition number is 597![Image](https://github.com/user-attachments/assets/11a55a06-fa38-49db-93bf-cea9cf4921cb)
so maybe we need a precondition before scalapack and cusolver too? I know little about numerical computing and would like to hear your advice. @kirk0830

**Comment by kirk0830** [comment]
@dzzz2001 hmmm...I think this condnum is not large (far smaller than those dangerous like 1e6), there may be other fatal problem. Could you upload your example here? I will also test it.

**Comment by jinzx10** [comment]
Hi @dzzz2001 , could you provide your test case (INPUT, STRU, KPT, pseudopotential & orbital file)?

**Comment by kirk0830** [comment]
@jinzx10 He has sent files to me, I will send a copy to you.

**Comment by jinzx10** [comment]
This issue seems to arise from the inconsistency in the orbital file. From the input files provided by @kirk0830 , the header of the orbital file shows lmax = 1, with one s orbital and one p orbital. However, the content of the file is 2s2p1d. I tried to change the header to lmax=2 and 2s2p1d, and abacus with scalapack solver can converge within 7 steps:
![Image](https://github.com/user-attachments/assets/5b46ffa0-1ebf-4c71-862a-509557c06a04)

Another thing I tried is to remove the extra orbitals in the orbital file to make it a consistent 1s1p orbital file. Again, abacus with scalapack can converge the calculation in 7 steps.
![Image](https://github.com/user-attachments/assets/07aa1e96-a241-48e6-b8af-e72741c35e8b)

I think what this issue reveals is that, when the content of the orbital file is inconsistent with its header, the calculation would have unexpected problems. One possible solution is to add some "consistency check". Alternatively, we may choose to support such inconsistency, as long as the number of orbitals listed in the header is less than the content. The choice may need wider discussion.
